edge/pkg/devicetwin/dtmanager/twin_test.go

```diff
@@ -0,0 +1,1604 @@
+}
+
+func (tw TwinWorker) sendMsg(msg *dttype.DTMessage, msgheart string, actionType string, contentType interface{}) {
+	if tw.ReceiverChan != nil {
+		msg.Action = actionType
+		msg.Msg.Content = contentType
+		tw.ReceiverChan <- msg
+	}
+	if tw.HeartBeatChan != nil {
+		tw.HeartBeatChan <- msgheart
+	}
+}
+
+func recMsg(channel chan interface{}, message *dttype.DTMessage) {
+	msg, _ := <-channel
+	*message = *msg.(*dttype.DTMessage)
+	return
+}
+
+func expectedFunc() *dttype.TwinValue {
+	var expected dttype.TwinValue
+	value := "value"
+	valueMetaData := &dttype.ValueMetadata{Timestamp: time.Now().UnixNano() / 1e6}
+	expected.Value = &value
+	expected.Metadata = valueMetaData
+	return &expected
+}
+
+func actualFunc() *dttype.TwinValue {
+	var actual dttype.TwinValue
+	value := "value"
+	valueMetaData := &dttype.ValueMetadata{Timestamp: time.Now().UnixNano() / 1e6}
+	actual.Value = &value
+	actual.Metadata = valueMetaData
+	return &actual
+}
+
+func keyTwinUpdateFunc() dttype.DeviceTwinUpdate {
+	var keyTwinUpdate dttype.DeviceTwinUpdate
+	twinKey := make(map[string]*dttype.MsgTwin)
+	twinKey["key1"] = &dttype.MsgTwin{
+		Expected: expectedFunc(),
+		Actual:   actualFunc(),
+		Metadata: &dttype.TypeMetadata{"nil"},
+	}
+	keyTwinUpdate.Twin = twinKey
+	keyTwinUpdate.BaseMessage = dttype.BaseMessage{EventID: "Event1"}
+	return keyTwinUpdate
+}
+
+func twinWorkerFunc(receiverChannel chan interface{}, confirmChannel chan interface{}, heartBeatChannel chan interface{}, context dtcontext.DTContext, group string) TwinWorker {
+	return TwinWorker{
+		Worker{
+			receiverChannel,
+			confirmChannel,
+			heartBeatChannel,
+			&context,
+		},
+		group,
+	}
+}
+
+func contextFunc() dtcontext.DTContext {
+	context := dtcontext.DTContext{
+		DeviceList:  &sync.Map{},
+		DeviceMutex: &sync.Map{},
+		Mutex:       &sync.Mutex{},
+	}
+	var testMutex sync.Mutex
+	context.DeviceMutex.Store("DeviceB", &testMutex)
+	var device dttype.Device
+	context.DeviceList.Store("DeviceB", &device)
+	return context
+}
+
+func msgFunc(content interface{}) interface{} {
+	return &model.Message{
+		model.MessageHeader{},
+		model.MessageRoute{},
+		content,
+	}
+}
+func msgFunc1(content interface{}) interface{} {
+	return model.Message{
+		model.MessageHeader{},
+		model.MessageRoute{},
+		content,
+	}
+}
+
+func TestStart(t *testing.T) {
+	keyTwinUpdate := keyTwinUpdateFunc()
+	content, _ := json.Marshal(keyTwinUpdate)
+
+	commChan := make(map[string]chan interface{})
+	channel := make(chan interface{})
+	commChan[dtcommon.CommModule] = channel
+
+	context := dtcontext.DTContext{
+		DeviceList:    &sync.Map{},
+		DeviceMutex:   &sync.Map{},
+		Mutex:         &sync.Mutex{},
+		CommChan:      commChan,
+		ModulesHealth: &sync.Map{},
+	}
+	var testMutex sync.Mutex
+	context.DeviceMutex.Store("DeviceB", &testMutex)
+	msgAttr := make(map[string]*dttype.MsgAttr)
+	device := dttype.Device{
+		ID:          "id1",
+		Name:        "DeviceB",
+		Description: "",
```

Comment:
default value of string type is "", no need  for explicit assignment here.
Applicable to similar places below.